src/plots/plots.js

```diff
@@ -1039,9 +1039,21 @@ plots.supplyTransformDefaults = function(traceIn, traceOut, layout) {
             _module = transformsRegistry[type],
             transformOut;
 
+        /*
+         * Supply defaults may run twice. First pass runs all supply defaults steps
+         * and adds the _module to any output transforms.
+         * If transforms exist another pass is run so that any generated traces also
+         * go through supply defaults. This has the effect of rerunning
+         * supplyTransforms. If the transform does not have a `transform` function
+         * it could not have generated any new traces and the second stage is
+         * unnecessary. We detect this case with the following variables.
+         */
+        var isSecondStage = transformIn._module && transformIn._module === _module,
```

Comment:
Is a third stage possible if multiple transforms expand things multiple times? (e.g. groupby + groupby) I think this makes sense to me, but just making sure transform module identity equates to `isSecondStage`. (Or is it `isNotFirstStage`?) My understanding is just a little loose here.

Reply:
right, more than 2 stages... need to think about that. Suggestions?

Reply:
Right I think we are ok. I should call this variable `isNotFirstStage` though. Thanks

Reply:
Things get fundamentally a bit shaky after one level. Like `groupby` + `groupby`. It does mostly what you'd expect, but it turns out *styling* is the big challenge. Think SQL. You process the data and process the data until you have a nice query that spits out data. Then you insert it into a plot and you're happy. Plotly's transforms incrementally apply styles *as* as sequence of transforms is applied so that you'd have to track styles as well as the pathway a data point took through the query to really sort things out.

Reply:
Sorry, specifics: `groupby` adds a `_group` property to traces so that you can figure out which group split generated a trace. But it should really be tracking that as a transform index and a group value. This comes into play with transform vs. calcTransform since you start having to worry about *when* a trace was split/grouped/styled/etc.

Reply:
Apologies if that's still a bit loose on the specifics. /cc @alexcjohnson since we've talked this through before in a fair bit of detail. I think our conclusion at the time was that the basic use cases are handled just fine and account for 97+% of realistic use-cases.